### ISSUE TITLE
Add an indexed prefix option to CoRI

### ISSUE THREAD

**Issue by jimsch** [core-wg/coral#46]
I think we should consider adding a prefixed option in parallel with the path type option.  This option would take an index into an array of possible prefixes.  The prefixes could be limited to absolute URIs with scheme and authority information or could allow for just a path type/path set of information.

Doing this would allow for the binary encoded CoRAL documents to match closer to the text ones in terms of how the base option is used.  That is there could be a single default base option which represents the base of the document and then a set of other prefix options which could encode the beginning of a URI that could then be tagged as a prefix if it is used multiple times in a CoRAL document.

**Comment by ektrah** [comment]
Currently in the binary format, link relation types are pretty much just text strings that just happen to conform to the syntax of IRIs. The idea here was to compress them only in their entirety with a static dictionary, so that implementations wouldn’t even need to decompress them and could just refer to them using their dictionary keys. (In the textual format, link relation types can be written shorter using `#using`.)

The link targets point at resources on the Web and are encoded as CoRIs. The assumption here so far is that link targets that appear next to each other in a document are also likely to share a common base, which can be factored out for compression. These need to be decompressed when the client wants to make a request to the link target. (In the textual format, link targets can be written shorter using `#base`.)

Which do you suggest to optimize with your proposal?

**Comment by jimsch** [comment]
I don't think that the first paragraph is relevant as I was looking at changing CoRIs and not dictionaries.  There might be benefits to changing that at some point, but I am not looking at that yet.

I am looking at the link targets.  So you could have a document that looks something like:

~~~
[ 
   [ 9, 1, [1, "coaps", 2, "host", 4, 5683]],
   [ 9, 2, [1, "coap", 2, "host", 4, 5683]],
   [ 9, 3, [1, "http", 2, "host", 4, 80]],
   [ 2, "rel1", [5, 0, "path1]],
   [ 2, "rel2", [TBD, 2, "path2"]]
]
~~~

This makes a difference if you are doing alternating between a couple of different bases - for example between a coap and a coaps scheme or you are using the same authority information in multiple target attributes (for example pointing to the same ACE authority server multiple times).

**Comment by ektrah** [comment]
Got it now. How do we keep the space requirements low for implementations that do not have enough memory to store all these bases? Bases can currently be defined at each level of depth, not just at the top level. But since there can be only one base, an implementation only needs to have enough memory for one base per level of depth. I’m worried that this might be already too much. 

Could you craft a bunch of non-toy example  documents that exhibit these characteristics (alternating between bases; pointing at ACE servers) and share them?

**Comment by jimsch** [comment]
Question - Are you assuming that the device can hold the entire CoRAL document in memory?  If so then keeping the multiple base definitions is a delta on the document.

**Comment by ektrah** [comment]
So far, nothing prevents a memory-constrained implementation from parsing the bytes of a CoRAL CBOR representation incrementally, discarding all the parsed data items immediately, until it has found the element it’s looking for. For that, the implementation only needs a fixed-sized buffer to hold chunks of the representation, plus some fixed-size memory for the current context and current base for each level of nesting (up to some specific implementation limit, like 2 or 3 levels), plus some memory for the currently decoded element.

**Comment by chrysn** [comment]
I'd suggest viewing this issue as a special case of "Can a document provide an in-line (additional?) dictionary?".

Jim's example could then be expressed as

```
[ 
   [ TBD-inline-dictionary,
      {1: [1, "coaps", 2, "host", 4, 5683]]},
      {2: [1, "coap", 2, "host", 4, 5683]]},
      {3: [1, "http", 2, "host", 4, 80]]},
   ],    
   [ 2, "rel1", [5, 0, "path1]],
   [1, TBD6(3)],
   [ 2, "rel2", [5, 0, "path2"]]
]
```
which uses a base directive to pick a base URI from the dictionary before expressing a link target.

This won't help with the question of "how does that fit into the constraints of a device that streams CoRAL files", but may help finding the non-toy examples ektrah asked about.

**Comment by ektrah** [comment]
Security-related issues to inline dictionaries:
* If we first compress and then encrypt a CoRAL document, then it's that information is leaked about what gets compressed and what doesn't.
* If the compression is too efficient, a small CoRAL document might need an excessive amount of time for decompression
